File: pt5/1181.before-and-after-puzzle.go
package main

import (
	"sort"
	"strings"
)

func beforeAndAfterPuzzles(phrases []string) []string {
	words := make([][]string, len(phrases))
	for i, phrase := range phrases {
		words[i] = strings.Split(phrase, " ")
	}

	puzzles := map[string]bool{}
	for i, firstWords := range words {
		for j, secondWords := range words {
			if i == j || firstWords[len(firstWords)-1] != secondWords[0] {
				continue
			}
			tail := strings.Join(secondWords[1:], " ")
			if tail == "" {
				puzzles[phrases[i]] = true
			} else {
				puzzles[phrases[i]+" "+tail] = true
			}
		}
	}

	answer := make([]string, 0, len(puzzles))
	for puzzle := range puzzles {
		answer = append(answer, puzzle)
	}
	sort.Strings(answer)
	return answer
}

File: pt5/1181.before-and-after-puzzle.py
from __future__ import annotations

from typing import List


class Solution:
    def beforeAndAfterPuzzles(self, phrases: List[str]) -> List[str]:
        words = [phrase.split() for phrase in phrases]
        puzzles = set()

        for i, first_words in enumerate(words):
            for j, second_words in enumerate(words):
                if i == j:
                    continue
                if first_words[-1] == second_words[0]:
                    tail = " ".join(second_words[1:])
                    puzzles.add(phrases[i] if not tail else phrases[i] + " " + tail)

        return sorted(puzzles)

